docs: updating tutorial for COMBINE25 (#421)
Reed-CompBio/spras: docs/tutorial/intermediate.rst

- Enable post-analysis steps to generate post analysis information

Step 1: Transforming high throughput experimental data into SPRAS compatible input data
========================================================================================

1.1 What is the SPRAS-standardized input data?
-----------------------------------------------

A pathway reconstruction algorithm requires a set of input nodes and an interactome; however, each algorithm expects its inputs to follow a unique format.

To simplify this process, SPRAS requires all input data in a dataset to be formatted once into a standardized SPRAS format.
SPRAS then automatically generates algorithm-specific input files when an algorithm is enabled in the configuration file.

.. note::
    Each algorithm uses the input nodes to guide or constrain the optimization process used to construct reconstruct subnetworks.

    An algorithm maps these input nodes onto the interactome and uses the network to identify connecting paths between the input nodes to form subnetworks.

Pathway reconstruction algorithms differ in the inputs nodes they require and how they interpret those nodes to identify subnetworks.

> - Some use source and target nodes to defined start and end points. 

Reviewer: Should we refer to the algorithm docs as a place to describe this? We'll have that after #408.

Author: Yes, if we push #408 before the tutorial, I will link to that part of the docs for further information on this.

- Some use prizes, which assign numerical scores assigned to nodes of interest.
- Some rely on active nodes, representing nodes that are significantly “on” under specific conditions.

An example of a node file required by SPRAS follows a tab-separated format:

.. code-block:: text

    NODEID  prize sources targets active
    A       1.0     True            True
    B       3.3             True    True
    C       2.5     True            True
    D       1.9             True    True

.. note::
    If a user provides only one type of input node but wants to run algorithms that require a different type, SPRAS can automatically convert the inputs into the compatible format:

    - Source-target nodes can be used with all algorithms by making a prize column set to 1 and an active column set to True.
    - Prize data can be adapted for active based algorithms by automatically making an active column set to True.
    - Active data can be adapted for prize based algorithms by making a prize column set to 1.

Along with differences in their inputs nodes, pathway reconstruction algorithms also interpret the input interactome differently.

> - Some can handle directed interactomes, others work only with undirected interactomes, and a few support mixed directionaltiy interactomes.

Reviewer: ```suggestion
- Some can handle directed interactomes, others work only with undirected interactomes, and a few support mixed directionality interactomes.
```
Do we need to define directionality of interactomes and mixed graphs?

SPRAS automatically converts the user-provided edge file into the format expected by each algorithm, ensuring that the directionality of the interactome matches the algorithm's requirements.

An example of an edge file required by SPRAS follows a tab-separated format:

.. code-block:: text

    A	B   0.98   U
    B	C   0.77   D

.. note::  
    SPRAS supports multiple standardized input formats.
    More information about input data formats can be found in the ``inputs/README.md`` file within the SPRAS repository.

1.2 Example high throughput data
---------------------------------

An example dataset is the EGF response mass spectrometry data, which includes three biological replicates, each with two technical replicates (IMAC and IP).
For each biological replicate, the technical replicates are combined to produce a single dataset of peptide abundances measured over time (0-124 minutes).

.. note::
    Mass spectrometry is a technique used to measure and identify proteins in a sample.

>     It works by breaking proteins into smaller pieces called peptides and measuring their mass using their mass-to-charge ratio.

Reviewer: ```suggestion
    It works by breaking proteins into smaller pieces called peptides and measuring their mass-to-charge ratio, which enables identifying which peptide is being measured.
```

    The data show how much of each peptide is present which can show protein levels change under different conditions.

>     Since proteins interact with each other in biological pathways, changes in their levels can reveal which parts of a pathway are active or affected.

Author: No I think we should at least add in phosphorylation abundances. I think it is important since people might try to use this approach with out mass spec data.

.. list-table::
   :header-rows: 1
   :widths: 20 15 10 10 10 10 10 10 10 10 10 10 10

   * - peptide
     - protein
     - gene.name
     - modified.sites
     - 0 min
     - 2 min
     - 4 min
     - 8 min
     - 16 min
     - 32 min
     - 64 min
     - 128 mn
     - replicate
   * - K.n[305.21]AFWMAIGGDRDEIEGLS[167.00]S[167.00]DEEH.-
     - Q6PD74,B4DG44,Q5JPJ4,Q6AWA0
     - AAGAB
     - S310,S311
     - 2.17
     - 2.09
     - 1.98
     - 1.78
     - 1.99
     - 2.12
     - 2.25
     - 1.46
     - C
   * - K.n[305.21]AFWMAIGGDRDEIEGLS[167.00]S[167.00]DEEH.-
     - Q6PD74,B4DG44,Q5JPJ4,Q6AWA0
     - AAGAB
     - S310,S311
     - 4.03
     - 3.73
     - 3.32
     - 3.36
     - 3.35
     - 3.37
     - 3.35
     - 3.86
     - B
   * - K.n[305.21]AFWMAIGGDRDEIEGLS[167.00]S[167.00]DEEH.-
     - Q6PD74,B4DG44,Q5JPJ4,Q6AWA0
     - AAGAB
     - S310,S311
     - 5.60
     - 4.75
     - 4.69
     - 4.59
     - 4.32
     - 4.90
     - 4.90
     - 5.48
     - A
     
1.4 Computing p-values using Tukey's HSD Test
-----------------------------------------------

We want to calculate the p-values per peptide.
This tells us how likely changes in abundance happen by chance.

> We use Tukey's Honest Significant Difference (HSD) test to compare all time points and correct for multiple testing to get a p-value for every pair of time points.

Author: The original values; I got them from the data they provided.

.. list-table::
   :header-rows: 1
   :widths: 25 20 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7

   * - peptide
     - protein
     - 2min vs 0min
     - 4min vs 0min
     - 8min vs 0min
     - 16min vs 0min
     - 32min.vs.0min
     - 64min.vs.0min
     - 128min.vs.0min
     - 4min.vs.2min
     - 8min.vs.2min
     - 16min.vs.2min
     - 32min.vs.2min
     - 64min.vs.2min
     - 128min.vs.2min
     - 8min.vs.4min
     - 16min.vs.4min
     - 32min.vs.4min
     - 64min.vs.4min
     - 128min.vs.4min
     - 16min.vs.8min
     - 32min.vs.8min
     - 64min.vs.8min
     - 128min.vs.8min
     - 32min.vs.16min
     - 64min.vs.16min
     - 128min.vs.16min
     - 64min.vs.32min
     - 128min.vs.32min
     - 128min.vs.64min
   * - K.n[305.21]ADVLEAHEAEAEEPEAGK[432.30]S[167.00]EAEDDEDEVDDLPSSR.R
     - QQ6PD74,B4DG44,Q5JPJ4,Q6AWA0
     - 0.67
     - 0.25
     - 0.14
     - 0.12	
     - 0.52
     - 0.76
     - 0.84
     - 0.99
     - 0.93	
     - 0.90	
     - 1.00	
     - 1.00
     - 1.00
     - 1.00
     - 1.00
     - 1.00
     - 0.97
     - 0.94
     - 1.00
     - 0.98
     - 0.87
     - 0.80
     - 0.96
     - 0.83
     - 0.75
     - 1.00
     - 1.00
     - 1.00

> Lower p-values indicate stronger evidence that a peptide's abundance significantly changes between those time points.

Author: If you could rephase this yes, I'm not sure how to phrase it correctly. 

Otherwise my attempt at a rephrase is: 

Lower p-values indicate that the observed differences in peptide abundance between time points are less likely to have occurred by chance, suggesting stronger evidence against the null hypothesis.

1.5 From p-values to prizes 
----------------------------

> P-values are transformed using -log10(p-value) so smaller p-values give larger prize scores.

Reviewer 2: ```suggestion
P-values are transformed using ``-log10(p-value)`` so smaller p-values give larger prize scores.
```

For each protein mapped to multiple peptides, the maximum prize value across all its peptides is assigned.

Finally, all protein identifiers (using the first one listed for each protein) are converted to UniProt Entry Names to match the identifiers that will be used in the interactome. 

.. note::
    All node identifiers should use the same namespace across every part of the data in a dataset.

.. list-table::
   :header-rows: 1
   :widths: 25 20 7 7 7

   * - peptide
     - protein
     - uniprot entry name
     - min p-value
     - -log10(min p-value)
   * - K.n[305.21]AFWMAIGGDRDEIEGLS[167.00]S[167.00]DEEH.-
     - Q6PD74,B4DG44,Q5JPJ4,Q6AWA0
     - AAGAB_HUMAN 
     - 0.12392034609392
     - 0.906857382317364

Input node data put into a SPRAS-standardized format:

.. code-block:: text

    NODE_ID     prize
    AAGAB_HUMAN	0.906857382

1.6 From Prizes to Source and Targets / Actives 
-----------------------------------------------

.. image:: ../_static/images/egf-pathway.png
   :alt: description of the image
   :width: 400
   :align: center

.. raw:: html

   <div style="margin:20px 0;"></div>

> Using known pathway knowledge [1]_ [2]_ [3]_:

Author: It mentioned on the site when referencing any pathway to include all three links. 

https://www.genome.jp/kegg/kegg1.html

> Please cite the following article(s) when using KEGG.
Kanehisa, M., Furumichi, M., Sato, Y., Matsuura, Y. and Ishiguro-Watanabe, M.; KEGG: biological systems database as a model of the real world. Nucleic Acids Res. 53, D672-D677 (2025). [[pubmed](https://www.ncbi.nlm.nih.gov/pubmed/39417505)] [[doi](https://doi.org/10.1093/nar/gkae909)]
Kanehisa, M; Toward understanding the origin and evolution of cellular organisms. Protein Sci. 28, 1947-1951 (2019) [[pubmed](https://pubmed.ncbi.nlm.nih.gov/31441146)] [[doi](https://doi.org/10.1002/pro.3715)]
Kanehisa, M. and Goto, S.; KEGG: Kyoto Encyclopedia of Genes and Genomes. Nucleic Acids Res. 28, 27-30 (2000). [[pubmed](https://pubmed.ncbi.nlm.nih.gov/10592173)] [[doi](https://doi.org/10.1093/nar/28.1.27)]

> - EGF serves as a source for the pathway.

Reviewer: ```suggestion
- EGF serves as a source for the pathway and was the experimental treatment.
```

- EGF is known to initiate signaling, so it can be assigned a high score (greater than all other nodes) to emphasize its importance and guide algorithms to start reconstruction from this point.

> - EGFR acts as a target in the pathway.

Reviewer: EGFR comes from the data, right? Not prior knowledge?

- All other downstream proteins detected in the data can also treated as targets
- All proteins in the data can be considered active since they correspond to proteins that are active under the given biological condition

Input node data put into a SPRAS-standardized format:

.. code-block:: text

    NODE_ID     prize       source  target  active
    AAGAB_HUMAN	0.906857382         True   True
    ... more nodes
    EGF_HUMAN	10	        True	True	True
    EGFR_HUMAN	6.787874699		    True	True
    ... more nodes

1.8 Finding an Interactome to use
----------------------------------

> To connect our proteins, we use a background PPI network, the interactome.

Reviewer 2: We should preserve the acronym expansion.

```suggestion
To connect our proteins, we use a background protein-protein interaction (PPI) network, or the interactome.
```

   │   └── log/
   │       └── ... snakemake log files ...
   ├── config/
   │   ├── basic.yaml
   │   ├── intermediate.yaml
   │   └── ... other configs ...
   ├── inputs/
   │   ├── phosphosite-irefindex13.0-uniprot.txt # pre-defined in SPRAS already, used by the intermediate.yaml file
   │   ├── tps-egfr-prizes.txt # pre-defined in SPRAS already, used by the intermediate.yaml file
   │   └── ... other input data ...
   ├── outputs/
   │   └── basic/
   │       └── ... output files ...

2.1 Algorithms in SPRAS
---------------------------------

SPRAS supports a wide range of algorithms, each designed around different biological assumptions and optimization strategies 
(See :doc:`Pathway Reconstruction Methods <../prms/prms>` for SPRAS's list of integrated algorithms.)

Wrapped algorithms
^^^^^^^^^^^^^^^^^^^

> Each pathway reconstruction algorithm within SPRAS has actually been wrapped by SPRAS.

Reviewer: ```suggestion
Each pathway reconstruction algorithm within SPRAS has been wrapped for SPRAS, meaning it has been prepared for the SPRAS framework.
```
